File: app/src/main/java/com/absinthe/anywhere_/adapter/page/PageTitleProvider.java
package com.absinthe.anywhere_.adapter.page;

import android.animation.ObjectAnimator;
import android.view.View;
import android.widget.ImageView;

import androidx.appcompat.view.menu.MenuBuilder;
import androidx.appcompat.widget.PopupMenu;

import com.absinthe.anywhere_.AnywhereApplication;
import com.absinthe.anywhere_.R;
import com.absinthe.anywhere_.model.AnywhereEntity;
import com.absinthe.anywhere_.model.GlobalValues;
import com.absinthe.anywhere_.model.PageEntity;
import com.absinthe.anywhere_.utils.manager.ActivityStackManager;
import com.absinthe.anywhere_.utils.manager.DialogManager;
import com.chad.library.adapter.base.entity.node.BaseNode;
import com.chad.library.adapter.base.provider.BaseNodeProvider;
import com.chad.library.adapter.base.viewholder.BaseViewHolder;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;

public class PageTitleProvider extends BaseNodeProvider {

    public static boolean isEditMode = false;

    @Override
    public int getItemViewType() {
        return 1;
    }

    @Override
    public int getLayoutId() {
        return R.layout.item_page_title;
    }

    @Override
    public void convert(@NotNull BaseViewHolder baseViewHolder, @Nullable BaseNode baseNode) {
        PageTitleNode node = (PageTitleNode) baseNode;
        if (node != null) {
            baseViewHolder.setText(R.id.tv_title, node.getTitle());
            ImageView ivArrow = baseViewHolder.getView(R.id.iv_arrow);
            if (node.isExpanded()) {
                onExpansionToggled(ivArrow, true);
            }
        }

    }

    @Override
    public void onClick(@NotNull BaseViewHolder helper, @NotNull View view, BaseNode data, int position) {
        if (isEditMode) {
            return;
        }

        Objects.requireNonNull(
                getAdapter()).expandOrCollapse(position);

        PageTitleNode node = (PageTitleNode) data;
        if (node != null) {
            ImageView ivArrow = helper.getView(R.id.iv_arrow);
            if (node.isExpanded()) {
                onExpansionToggled(ivArrow, true);
            } else {
                onExpansionToggled(ivArrow, false);
            }
        }
    }

    @Override
    public boolean onLongClick(@NotNull BaseViewHolder helper, @NotNull View view, BaseNode data, int position) {
        if (isEditMode) {
            return false;
        }

        PopupMenu popup = new PopupMenu(getContext(), view);
        popup.getMenuInflater()
                .inflate(R.menu.page_menu, popup.getMenu());
        if (popup.getMenu() instanceof MenuBuilder) {
            MenuBuilder menuBuilder = (MenuBuilder) popup.getMenu();
            menuBuilder.setOptionalIconsVisible(true);
        }

        PageTitleNode node = (PageTitleNode) data;
        if (node != null) {
            popup.setOnMenuItemClickListener(item -> {
                switch (item.getItemId()) {
                    case R.id.rename_page:
                        DialogManager.showRenameDialog(ActivityStackManager.getInstance().getTopActivity(), node.getTitle());
                        break;
                    case R.id.delete_page:
                        DialogManager.showDeletePageDialog(getContext(), node.getTitle(), (dialog1, which) -> {
                            AnywhereApplication.sRepository.deletePage(getPageEntity(node.getTitle()));
                            List<PageEntity> list = AnywhereApplication.sRepository.getAllPageEntities().getValue();
                            if (list != null) {
                                String title = list.get(0).getTitle();
                                List<AnywhereEntity> anywhereEntities = AnywhereApplication.sRepository.getAllAnywhereEntities().getValue();
                                if (anywhereEntities != null) {
                                    for (AnywhereEntity ae : anywhereEntities) {
                                        if (ae.getCategory().equals(node.getTitle())) {
                                            ae.setCategory(title);
                                            AnywhereApplication.sRepository.update(ae);
                                        }
                                    }
                                }
                                GlobalValues.setsCategory(title, 0);
                            }
                        }, false);
                        break;
                    case R.id.delete_page_and_item:
                        DialogManager.showDeletePageDialog(getContext(), node.getTitle(), (dialog1, which) -> {
                            AnywhereApplication.sRepository.deletePage(getPageEntity(node.getTitle()));
                            List<PageEntity> list = AnywhereApplication.sRepository.getAllPageEntities().getValue();
                            if (list != null) {
                                List<AnywhereEntity> anywhereEntities = AnywhereApplication.sRepository.getAllAnywhereEntities().getValue();
                                if (anywhereEntities != null) {
                                    for (AnywhereEntity ae : anywhereEntities) {
                                        if (ae.getCategory().equals(node.getTitle())) {
                                            AnywhereApplication.sRepository.delete(ae);
                                        }
                                    }
                                }
                                GlobalValues.setsCategory(list.get(0).getTitle(), 0);
                            }
                        }, true);
                        break;
                    default:
                }
                return true;
            });
        }

        popup.show();

        return super.onLongClick(helper, view, data, position);
    }

    private void onExpansionToggled(ImageView arrow, boolean expanded) {
        float start, target;
        if (expanded) {
            start = 0f;
            target = 90f;
        } else {
            start = 90f;
            target = 0f;
        }
        ObjectAnimator objectAnimator = ObjectAnimator.ofFloat(arrow, View.ROTATION, start, target);
        objectAnimator.setDuration(200);
        objectAnimator.start();
    }

    private static PageEntity getPageEntity(String title) {
        List<PageEntity> list = AnywhereApplication.sRepository.getAllPageEntities().getValue();
        if (list != null) {
            for (PageEntity pe : list) {
                if (pe.getTitle().equals(title)) {
                    return pe;
                }
            }
        }
        return null;
    }

    public static void renameTitle(String oldTitle, String newTitle) {
        PageEntity pe = getPageEntity(oldTitle);
        List<AnywhereEntity> list = AnywhereApplication.sRepository.getAllAnywhereEntities().getValue();
        if (list != null && pe != null) {
            for (AnywhereEntity ae : list) {
                if (ae.getCategory().equals(pe.getTitle())) {
                    ae.setCategory(newTitle);
                    AnywhereApplication.sRepository.update(ae);
                }
            }
            AnywhereApplication.sRepository.deletePage(pe);
            pe.setTitle(newTitle);
            AnywhereApplication.sRepository.insertPage(pe);
            GlobalValues.setsCategory(newTitle);
        }
    }
}

File: app/src/main/java/com/absinthe/anywhere_/adapter/page/PageTitleProvider.kt
package com.absinthe.anywhere_.adapter.page

import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.DialogInterface
import android.view.HapticFeedbackConstants
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import androidx.appcompat.view.menu.MenuBuilder
import androidx.appcompat.widget.PopupMenu
import com.absinthe.anywhere_.AnywhereApplication
import com.absinthe.anywhere_.R
import com.absinthe.anywhere_.constants.GlobalValues
import com.absinthe.anywhere_.model.database.PageEntity
import com.absinthe.anywhere_.utils.manager.ActivityStackManager
import com.absinthe.anywhere_.utils.manager.DialogManager
import com.chad.library.adapter.base.entity.node.BaseNode
import com.chad.library.adapter.base.provider.BaseNodeProvider
import com.chad.library.adapter.base.viewholder.BaseViewHolder

class PageTitleProvider : BaseNodeProvider() {

    override val itemViewType: Int
        get() = 1

    override val layoutId: Int
        get() = R.layout.item_page_title

    override fun convert(helper: BaseViewHolder, item: BaseNode) {
        val node = item as PageTitleNode

        helper.setText(R.id.tv_title, node.title)

        val ivArrow = helper.getView<ImageView>(R.id.iv_arrow)
        if (node.isExpanded) {
            onExpansionToggled(ivArrow, true)
        }
    }

    override fun onClick(helper: BaseViewHolder, view: View, data: BaseNode, position: Int) {
        if (isEditMode) {
            return
        }
        getAdapter()?.expandOrCollapse(position)
        val node = data as PageTitleNode
        val ivArrow = helper.getView<ImageView>(R.id.iv_arrow)

        if (node.isExpanded) {
            onExpansionToggled(ivArrow, true)
        } else {
            onExpansionToggled(ivArrow, false)
        }
    }

    @SuppressLint("RestrictedApi")
    override fun onLongClick(helper: BaseViewHolder, view: View, data: BaseNode, position: Int): Boolean {
        if (isEditMode) {
            return false
        }
        val popup = PopupMenu(context, view).apply {
            menuInflater.inflate(R.menu.page_menu, menu)

            if (menu is MenuBuilder) {
                val menuBuilder = menu as MenuBuilder
                menuBuilder.setOptionalIconsVisible(true)
            }
        }

        val node = data as PageTitleNode
        popup.setOnMenuItemClickListener { item: MenuItem ->
            when (item.itemId) {
                R.id.rename_page -> ActivityStackManager.topActivity?.let {
                    DialogManager.showRenameDialog(it, node.title)
                }
                R.id.delete_page -> DialogManager.showDeletePageDialog(context, node.title, DialogInterface.OnClickListener { _, _ ->
                    getPageEntity(node.title)?.let { AnywhereApplication.sRepository.deletePage(it) }
                    AnywhereApplication.sRepository.allPageEntities.value?.let { list ->
                        val title = list[0].title

                        AnywhereApplication.sRepository.allAnywhereEntities.value?.let {
                            for (ae in it) {
                                if (ae.category == node.title) {
                                    ae.category = title
                                    AnywhereApplication.sRepository.update(ae)
                                }
                            }
                        }

                        GlobalValues.setsCategory(title, 0)
                    }
                }, false)
                R.id.delete_page_and_item -> DialogManager.showDeletePageDialog(context, node.title, DialogInterface.OnClickListener { _: DialogInterface?, _: Int ->
                    getPageEntity(node.title)?.let { AnywhereApplication.sRepository.deletePage(it) }
                    AnywhereApplication.sRepository.allPageEntities.value?.let { list ->
                        AnywhereApplication.sRepository.allAnywhereEntities.value?.let {
                            for (ae in it) {
                                if (ae.category == node.title) {
                                    AnywhereApplication.sRepository.delete(ae)
                                }
                            }
                        }

                        GlobalValues.setsCategory(list[0].title, 0)
                    }
                }, true)
            }
            true
        }
        popup.show()
        view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
        return super.onLongClick(helper, view, data, position)
    }

    private fun onExpansionToggled(arrow: ImageView, expanded: Boolean) {
        val start: Float
        val target: Float
        if (expanded) {
            start = 0f
            target = 90f
        } else {
            start = 90f
            target = 0f
        }

        ObjectAnimator.ofFloat(arrow, View.ROTATION, start, target).apply {
            duration = 200
            start()
        }
    }

    companion object {
        @JvmField
        var isEditMode = false

        private fun getPageEntity(title: String): PageEntity? {
            AnywhereApplication.sRepository.allPageEntities.value?.let {
                for (pe in it) {
                    if (pe.title == title) {
                        return pe
                    }
                }
            }
            return null
        }

        @JvmStatic
        fun renameTitle(oldTitle: String, newTitle: String) {
            val pe = getPageEntity(oldTitle)
            AnywhereApplication.sRepository.allAnywhereEntities.value?.let { list ->
                pe?.let {
                    for (ae in list) {
                        if (ae.category == pe.title) {
                            ae.category = newTitle
                            AnywhereApplication.sRepository.update(ae)
                        }
                    }
                    AnywhereApplication.sRepository.deletePage(pe)
                    pe.title = newTitle
                    AnywhereApplication.sRepository.insertPage(pe)
                    GlobalValues.category = newTitle
                }
            }
        }
    }
}